articles/external_qos_file.md

```diff
@@ -0,0 +1,672 @@
+
+The first alternative is limited, as the proposed convention would use exactly the same profile in all publishers in a topic.
+This convention can be modified to match node name, topic name and profile id, but that approach will be hard to understand.
+The second approach leverages the new file format and is more powerful.
+
+As an example of the second alternative:
+
+```xml
+<qos_profiles>
+    <node name="my_node" namespace="/my_ns/nested_ns">
+        <publisher topic_name="asd">
+            <qos>
+                <rmw_payload>
+                    <profile_name>my_qos_profile_name</profile_name>
+                </rmw_payload>
+            </qos>
+        </publisher>
+    </node>
+</qos_profiles>
+```
+
+```yaml
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        publisher:
+            topic_name: asd
+            rmw_payload:
+                profile_name: my_qos_profile_name
+```
+
+For this to work, `rmw_qos_profile_t` will need to be extended with a `rmw_qos_profile_payload_t` member defined like:
+
+```c
+typedef struct rmw_qos_profile_payload_item_t {
+  char * key;
+  char * value;
+} rmw_qos_profile_payload_item_t;
+
+typedef struct rmw_qos_profile_payload_t {
+  rmw_qos_profile_payload_item_t * key_value_pairs;
+} rmw_qos_profile_payload_t;
+```
+
+It would be up to the implementation what to do with the passed key value pairs.
+
+Note:
+  The proposed API is limited in that it treats all value as strings instead of taking advantage of natural representations supported by some markup languages.
+
+## File format
+
+As showed in the above examples, both YAML or XML can correctly represent the desired file format.
+
+XML advantages:
+- Can be easily verified.
+
+YAML advantages:
+- It can be argued that the format is simpler and more readable than XML.
+- The format proposed in the above examples can easily be integrated with the current parameter file format.
+  That integration would lead in a unique "configuration file", that could be further extended in the future.
+  For example, passing remapping rules in it.
+
+### Author's opinion
+
+Using yaml sounds preferred.
+libyaml is already a dependency of `rcl`, and the proposed format can be easily merged with the existing parameter file format if that's desired in the future.
+
+Though through the document examples were showed both in XML and YAML, supporting both doesn't sound like a good idea.
+
+## Explicitly allowing external configurability in entities
+
+Nodes should have an option to enable having their QoS settings configured externally.
+This gives authors the ability to define the allowable range of behavior for their nodes.
+Furthermore, the option to externally configure QoS settings should be disabled by default (ie. node authors must opt-in).
+The rationale for an opt-in policy is to force node authors to have to think about the implications of allowing various QoS changes to the node entities.
```

Comment:
I was thinking opposite. It would be hard for node authors to make decision since this is dependent on what application requires. (i think that i am gonna find difficulty to make decision so maybe i am gonna just enables it.) even with the same `node`, `topics` and `services`, we'd like to adjust the QoS based on the use cases.

Reply:
I would vote for node author opt-in. Otherwise you can end up in out of spec situations with strange behavior that is maybe not obvious when debugging